packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts

```diff
@@ -0,0 +1,275 @@
+import {DevSessionLogger} from './dev-session-logger.js'
+import {AppEvent, AppEventWatcher} from '../../app-events/app-event-watcher.js'
+import {DevSessionStatusManager} from '../dev-session-status-manager.js'
+import {DevSessionProcessOptions} from '../dev-session.js'
+import {getExtensionUploadURL} from '../../../deploy/upload.js'
+import {outputDebug} from '@shopify/cli-kit/node/output'
+import {endHRTimeInMs, startHRTime} from '@shopify/cli-kit/node/hrtime'
+import {ClientError} from 'graphql-request'
+import {performActionWithRetryAfterRecovery} from '@shopify/cli-kit/common/retry'
+import {JsonMapType} from '@shopify/cli-kit/node/toml'
+import {AbortError} from '@shopify/cli-kit/node/error'
+import {isUnitTest} from '@shopify/cli-kit/node/context/local'
+import {dirname, joinPath} from '@shopify/cli-kit/node/path'
+import {readFileSync, writeFile} from '@shopify/cli-kit/node/fs'
+import {zip} from '@shopify/cli-kit/node/archiver'
+import {formData, fetch} from '@shopify/cli-kit/node/http'
+import {AbortSignal} from '@shopify/cli-kit/node/abort'
+import {Writable} from 'stream'
+
+interface DevSessionPayload {
+  shopFqdn: string
+  appId: string
+  assetsUrl: string
+}
+
+export interface UserError {
+  message: string
+  on: JsonMapType
+  field?: string[] | null
+  category: string
+}
+
+type DevSessionResult =
+  | {status: 'updated' | 'created' | 'aborted'}
+  | {status: 'remote-error'; error: UserError[]}
+  | {status: 'unknown-error'; error: Error}
+
+export class DevSession {
+  private readonly statusManager: DevSessionStatusManager
+  private readonly logger: DevSessionLogger
+  private readonly options: DevSessionProcessOptions
+  private readonly appWatcher: AppEventWatcher
+  private readonly stdout: Writable
+  private readonly bundlePath: string
+  private bundleControllers: AbortController[] = []
+
+  constructor(processOptions: DevSessionProcessOptions, stdout: Writable, _stderr: Writable, _signal: AbortSignal) {
+    this.statusManager = processOptions.devSessionStatusManager
+    this.logger = new DevSessionLogger(stdout)
+    this.options = processOptions
+    this.appWatcher = processOptions.appWatcher
+    this.stdout = stdout
+    this.bundlePath = processOptions.appWatcher.buildOutputPath
+  }
+
+  async start() {
```

Comment:
Could we look at a private constructor, and a `public static async start(<constructor args>): Promise<DevSession>` method? Given the only way this class should be used is to create it and start it in one go. So from a caller its like `const session = await DevSession.start(...args...)`